publish-externals/src/index.ts

```diff
@@ -2,35 +2,114 @@ import path from "path"
-      for (let publication of ExternalPublications) {
-        console.log(`[${publication.name}] ${frontMatter.slug} ... publishing ⏳`)
-        try {
-          await publication.handler(frontMatter, content)
-          console.log(`[${publication.name}] Success ${frontMatter.slug} ... succeeded ✅`)
-        } catch (error) {
-          errors.push(error)
-          console.error(`[${publication.name}] Failure ${frontMatter.slug} ... failed 💀`)
+  const addedFilesArg = process.argv[2]
+  const addedFiles = addedFilesArg ? addedFilesArg.split(" ") : []
+  const blogs: {[key: string]: Blog} = snapshot.blogs || {}
+
+  if (addedFiles.length === 0 && Object.keys(blogs).length === 0) {
+    console.log("No changes detected. Exiting.")
+    return
+  }
+
+  try {
+    for (let publication of ExternalPublications) {
+      for (const file of addedFiles) {
+        if (file.startsWith("blog/")) {
+          await publish(file, blogs, publication)
+        }
+      }
+
+      for (const slug in blogs) {
+        const file = blogs[slug].file
+        if (!addedFiles.includes(file)) {
+          const {content} = extractFrontMatterAndContent(path.join(__dirname, "../../", file))
+          const contentHash = createHash("sha256").update(content).digest("hex")
+
+          if (!blogs[slug].hash || blogs[slug].hash !== contentHash) {
+            await publish(file, blogs, publication)
+          }
         }
       }
     }
+  } finally {
+    await writeSnapshot(blogs)
   }
+}
+
+const publish = async (file: string, blogs: {[key: string]: Blog}, publication: ExternalPublication) => {
+  const filePath = path.join(__dirname, "../../", file)
+  const {frontMatter, content} = extractFrontMatterAndContent(filePath)
+  const platform = publication.name
+  const contentHash = createHash("sha256").update(content).digest("hex")
+  const slug = frontMatter.slug
+
+  console.log(`[${platform}] ${frontMatter.slug} ... publishing ⏳`)
+  try {
+    await publication.handler(frontMatter, content)
+
+    blogs[slug] = blogs[slug] || {
+      file,
+      platforms: {},
+    }
+    blogs[slug].platforms[platform] = {
+      published: true,
+      lastUpdatePublished: true,
+      lastSuccessfulPublishedAt: new Date().toUTCString(),
+    }
+    blogs[slug].hash = contentHash
+
+    console.log(`[${platform}] Success ${frontMatter.slug} ... succeeded ✅`)
+  } catch (error) {
+    console.error(`[${platform}] Failure ${frontMatter.slug} ... failed 💀`)
+
+    blogs[slug] = blogs[slug] || {
+      file,
+      platforms: {},
+    }
+    blogs[slug].platforms[platform] = {
+      published: blogs[slug].platforms[platform]?.published || false,
+      lastUpdatePublished: false,
+    }
+    console.log(blogs[slug])
+  }
+}
 
-  if (errors.length !== 0) {
-    console.error(errors)
-    throw new Error("Publishing failed because of one or more errors")
+const writeSnapshot = async (blogs: {[key: string]: Blog}) => {
+  try {
```

Comment:
Should print the snapshot before writing so we can manually recover from failures by looking at the logs.